src/Symfony/Component/Security/Core/Security.php

```diff
@@ -57,4 +66,39 @@ public function getToken(): ?TokenInterface
     {
         return $this->container->get('security.token_storage')->getToken();
     }
+
+    public function login(UserInterface $user, AuthenticatorInterface $authenticator = null, string $firewallName = null): void
+    {
+        $request = $this->container->get('request_stack')->getCurrentRequest();
+
+        if (null === $authenticator) {
```

Comment:
These could be refactored to private getters, something like:
```php
$request = $this->getRequest();
$authenticator = $authenticator ?: $this->getAuthenticator($firewallName ?: $this->getFirewallName($request));
```

Reply:
Done